### DISCUSSION TITLE
Recursive git submodules

### DISCUSSION THREAD

**Discussion by Josef-Kainz-Reichardt** [renovatebot/renovate#30189] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab (ghcr.io/renovatebot/renovate:37.429.0)

### Please tell us more about your question or problem

We have a rust repo which has dependency on a crate which is located in a sub-module of a sub-module. But the problem is not rust specific.

Example directory structure:

- main repo
  - `demo-app` (with dependency on `demo-lib`
  - sub1 (a git submodule containing again a git submodule)
    - sub2 (a git submodule containing `demo-lib`)

You can clone the whole repo tree with `git clone --recursive ...` or update the submodules with `git submodule update --recursive`.

We have already set the toggle `--clone-submodules` and this works for the renovate bot when he works directly on the repo `sub1` in this example. But it doesn't work for the main repo because the bot only loads one level ov submodules but he don't do it recursively.

I have created a reproduction repo on GitHub where you can see the same behaviour: https://github.com/Josef-Reichardt/renovate-bot-reproduction-recursive-submodules/  
And here you can find a pull request from the renovate bot where he shows an error because of the missing sub-submodule: https://github.com/Josef-Reichardt/renovate-bot-reproduction-recursive-submodules/pull/3

### Logs (if relevant)

<details><summary>Logs</summary>

```

Command failed: cargo update --config net.git-fetch-with-cli=true --manifest-path demo-app/Cargo.toml --workspace
error: failed to get `renovate-bot-reproduction-recursive-submodules-sub2` as a dependency of package `renovate-bot-reproduction-recursive-submodules v0.1.0 (/tmp/renovate/repos/github/Josef-Reichardt/renovate-bot-reproduction-recursive-submodules/demo-app)`

Caused by:
  failed to load source for dependency `renovate-bot-reproduction-recursive-submodules-sub2`

Caused by:
  Unable to update /tmp/renovate/repos/github/Josef-Reichardt/renovate-bot-reproduction-recursive-submodules/renovate-bot-reproduction-recursive-submodules-sub1/renovate-bot-reproduction-recursive-submodules-sub2

Caused by:
  failed to read `/tmp/renovate/repos/github/Josef-Reichardt/renovate-bot-reproduction-recursive-submodules/renovate-bot-reproduction-recursive-submodules-sub1/renovate-bot-reproduction-recursive-submodules-sub2/Cargo.toml`

Caused by:
  No such file or directory (os error 2)

```

</details>

**Comment by rarkins** [comment, marked as answer]
Renovate doesn't do a regular `clone` for performance optimization reasons. When doing a branch checkout it does attempt to recurse submodules though: https://github.com/renovatebot/renovate/blob/17eacd2779649032731a516678784b6716edbc66/lib/util/git/index.ts#L530-L534

I tried adding `--recurse` to this logic: https://github.com/renovatebot/renovate/blob/17eacd2779649032731a516678784b6716edbc66/lib/util/git/index.ts#L354-L364

But it didn't fix your reproduction repo.

I've converted this to an issue as I don't have time to look into it any further at the moment: https://github.com/renovatebot/renovate/issues/30294

If you are able to find the solution, a PR would be welcome

**Reply by Josef-Kainz-Reichardt** [reply]
Thanks for your response. I've privided a PR: https://github.com/renovatebot/renovate/pull/30361
